core/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -574,23 +575,53 @@ public long getLastSyncedGlobalCheckpoint() {
      * Snapshots the current transaction log allowing to safely iterate over the snapshot.
      * Snapshots are fixed in time and will not be updated with future operations.
      */
-    public Snapshot newSnapshot() {
+    public Snapshot newSnapshot() throws IOException {
         try (ReleasableLock ignored = readLock.acquire()) {
-            return newSnapshot(getMinFileGeneration());
+            return newSnapshotFromGen(getMinFileGeneration());
         }
     }
 
-    public Snapshot newSnapshot(long minGeneration) {
+    public Snapshot newSnapshotFromGen(long minGeneration) throws IOException {
         try (ReleasableLock ignored = readLock.acquire()) {
             ensureOpen();
             if (minGeneration < getMinFileGeneration()) {
                 throw new IllegalArgumentException("requested snapshot generation [" + minGeneration + "] is not available. " +
                     "Min referenced generation is [" + getMinFileGeneration() + "]");
             }
-            Snapshot[] snapshots = Stream.concat(readers.stream(), Stream.of(current))
+            TranslogSnapshot[] snapshots = Stream.concat(readers.stream(), Stream.of(current))
                     .filter(reader -> reader.getGeneration() >= minGeneration)
-                    .map(BaseTranslogReader::newSnapshot).toArray(Snapshot[]::new);
-            return new MultiSnapshot(snapshots);
+                    .map(BaseTranslogReader::newSnapshot).toArray(TranslogSnapshot[]::new);
+            return newMultiSnapshot(snapshots);
+        }
+    }
+
+    public Snapshot newSnapshotFromMinSeqNo(long minSeqNo) throws IOException {
+        try (ReleasableLock ignored = readLock.acquire()) {
+            ensureOpen();
+            TranslogSnapshot[] snapshots = readersAboveMinSeqNo(minSeqNo).map(BaseTranslogReader::newSnapshot)
+                .toArray(TranslogSnapshot[]::new);
+            return newMultiSnapshot(snapshots);
+        }
+    }
+
+    private Snapshot newMultiSnapshot(TranslogSnapshot[] snapshots) throws IOException {
+        final Closeable onClose;
+        if (snapshots.length == 0) {
+            onClose = () -> {};
+        } else {
+            assert Arrays.stream(snapshots).map(BaseTranslogReader::getGeneration).min(Long::compareTo).get()
```

Comment:
I don't think so? acquiring the min gen from the translog deletion policy will keep all the other ones around which is what we care about here?